google-cloud-core/src/main/java/com/google/cloud/ServiceOptions.java

```diff
@@ -394,8 +394,13 @@ protected static String getAppEngineProjectId() {
       String serviceAccountName = (String) method.invoke(appIdentityService);
       int indexOfAtSign = serviceAccountName.indexOf('@');
       return serviceAccountName.substring(0, indexOfAtSign);
-    } catch (Exception ignore) {
-      // return null if can't determine
+    } catch (Exception exception) {
+      if (System.getProperty("com.google.appengine.runtime.version") != null) {
+        // Could not resolve appengine classes under GAE environment.
+        throw new UnsupportedOperationException("Unable to resolve appengine-sdk classes. "
```

Comment:
I don't think this exception is quite the right one. The operation is supported in general, but there are missing prerequisites.